core/capabilities/remote/trigger_publisher.go

```diff
@@ -0,0 +1,192 @@
+package remote
+
+import (
+	"context"
+	"errors"
+	sync "sync"
+	"time"
+
+	commoncap "github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/pb"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink/v2/core/capabilities/remote/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+)
+
+// TriggerPublisher manages all external users of a local trigger capability.
+// Its responsibilities are:
+//  1. Manage trigger registrations from external nodes (receive, store, aggregate, expire).
+//  2. Send out events produced by a concrete trigger implementation.
+//
+// TriggerPublisher communicates with corresponding TriggerSubscribers on remote nodes.
+//
+// Expected usage by a concrete trigger implementation:
+//  1. Call GetActiveRegistrations() to refresh the list of active registrations
+//  2. Call SendTriggerEvent() for every new event (batched by subscribed workflow IDs)
```

Comment:
I'm a little confused by point 2 -- does that imply that remote triggers will have to know about the trigger publisher, since they'll need to call SendTriggerEvent?

Reply:
Yes. It feels reasonable because we should have an explicit config saying whether a capability is open to remote calls or not (e.g. a CRON trigger will likely never be).
Alternatively, we could also implement it in a transparent way where TriggerPublisher calls Capability API on a real capability and it replies via callback channels. The downside is that we lose opportunities to optimize, most notably batch across workflows. But let me think about it more.